### PR TITLE
Modularly load the rest api and the explorer

### PR DESCRIPTION
**Pull request by hacksparrow**
This PR updates the templates to **loopback 2** conventions, along with the updated name of the explorer, `loopback-component-explorer`.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "loopback-workspace",
-  "version": "3.17.1",
+  "version": "3.18.0",
```
Comment (raymondfeng) [inline package.json]: Please don't change the version as our release tool will do it automatically.

Comment (hacksparrow) [inline package.json]: That's great! Thanks for letting me know.

File: templates/components/api-server/template/server/server.js [hunk 2]
```diff
@@ -3,20 +3,23 @@ var boot = require('loopback-boot');
...
+
 app.start = function() {
   // start the web server
   return app.listen(function() {
     app.emit('started');
-    console.log('Web server listening at: %s', app.get('url'));
+    var baseUrl = app.get('url').replace(/\/$/, '');
+    console.log('Web server listening at: %s', baseUrl);
+    if (app.get('loopback-explorer')) {
+      var explorerPath = app.get('loopback-explorer').mountPath;
```
Comment (bajtos) [inline templates/components/api-server/template/server/server.js]: I think these two lines should use `loopback-component-explorer`?

File: templates/components/api-server/template/server/server.js [hunk 3]
```diff
@@ -3,20 +3,23 @@ var boot = require('loopback-boot');
...
 
-// Bootstrap the application, configure models, datasources and middleware.
-// Sub-apps like REST API are mounted via boot scripts.
-boot(app, __dirname, function(err) {
-  if (err) throw err;
-
-  // start the server if `$ node server.js`
-  if (require.main === module)
-    app.start();
-});
```
Comment (bajtos) [inline templates/components/api-server/template/server/server.js]: Please revert this change, as it is effectively removing support for async boot scripts.

File: templates/components/api-server/template/server/server.js [hunk 4]
```diff
@@ -3,20 +3,23 @@ var boot = require('loopback-boot');
 
 var app = module.exports = loopback();
 
+// boot scripts mount components like REST API
+boot(app, __dirname);
+
```
Comment (bajtos) [inline templates/components/api-server/template/server/server.js]: Please revert this change, as it is effectively removing support for async boot scripts.

File: templates/components/api-server/template/server/middleware.json [hunk 5]
```diff
@@ -19,8 +19,14 @@
...
   },
   "routes": {
+    "loopback#rest": {
+      "paths": ["${restApiRoot}"]
+    }
   },
   "files": {
+    "serve-static": {
+      "params": "$!../client"
+    }
```
Comment (bajtos) [inline templates/components/api-server/template/server/middleware.json]: I think a SPA client should not be a part of "API Server" template, please remove this middleware. IIRC, `slc loopback:example` is already adding this middleware entry.

**Comment by bajtos** [comment]
Hi @hacksparrow, thank you for the pull request! I found few issues (see above), could you please address them?

File: templates/components/api-server/template/server/server.js (hunk 2 above)
Comment (hacksparrow) [inline templates/components/api-server/template/server/server.js]: Err, yes!

File: templates/components/api-server/template/server/server.js (hunk 3 above)
Comment (hacksparrow) [inline templates/components/api-server/template/server/server.js]: Maybe add a test for async boot scripts? I ran `npm test` and it did not complain, I assumed it would be all good.

**Comment by hacksparrow** [comment]
@bajtos thanks a lot. I will be making the changes you suggested.

**Comment by hacksparrow** [comment]
@bajtos I just made the changes. Please take a look.

**Comment by cebor** [comment]
:+1: i have waited for this ...

File: templates/components/api-server/component.js [hunk 6]
```diff
@@ -16,10 +16,9 @@ template.package = {
     "loopback": "^2.22.0",
     "loopback-boot": "^2.6.5",
     "loopback-datasource-juggler": "^2.39.0",
-    "serve-favicon": "^2.0.1"
-  },
-  "optionalDependencies": {
-    "loopback-explorer": "^1.1.0"
+    "loopback-component-explorer": "^2.1.0",
+    "serve-favicon": "^2.0.1",
+    "serve-static": "^1.10.0"
```
Comment (cebor) [inline templates/components/api-server/component.js]: i think you missed some configuration for this middleware

see the example app: https://github.com/strongloop/loopback-example-app/pull/77/files#diff-264492e8c99bdecaf2b533ecdb34a2ebR27

Comment (hacksparrow) [inline templates/components/api-server/component.js]: Those middleware are now added using `middleware.json`.

Comment (cebor) [inline templates/components/api-server/component.js]: yes i know, but i the example has these new lines in it:

```
"files": {
    "serve-static": {
      "params": "$!../client"
    }
  },
```

https://github.com/strongloop/loopback-example-app/blob/master/server/middleware.json#L26

Comment (hacksparrow) [inline templates/components/api-server/component.js]: @bajtos suggested that it be removed from the template.
> I think a SPA client should not be a part of "API Server" template, please remove this middleware.

**Comment by bajtos** [comment]
@hacksparrow we have renamed loopback-explorer to loopback-component-explorer in the meantime, could you take that into account too?

See https://github.com/strongloop/loopback-example-app/pull/79

File: templates/components/api-server/component.js (hunk 6 above)
Comment (bajtos) [inline templates/components/api-server/component.js]: > Yes i know, but in the example app these new lines were added to it, shouldn't we also add these here?

No, we shouldn't. This template is generating an "API server", the example application is built on top of this template (but it's not the same!).

In particular, the static client is configured here: https://github.com/strongloop/generator-loopback/blob/01a38d519e4b3c8b63c10a3a84c43c1b24aefdaa/example/index.js#L401-L421

**Comment by hacksparrow** [comment]
@bajtos I thought I replaced all instances of `loopback-explorer` to `loopback-component-explorer` in this project? Or am I missing something?

**Comment by cebor** [comment]
I think he means the package.json in this project 😀

**Comment by cebor** [comment]
Ok forget what I said...

File: templates/components/api-server/component.js (hunk 6 above)
Comment (bajtos) [inline templates/components/api-server/component.js]: @hacksparrow please remove `serve-static` from package.json dependencies since it's not used.

**Comment by bajtos** [comment]
> I thought I replaced all instances of loopback-explorer to loopback-component-explorer in this project? Or am I missing something?

You are right, I must have missed that.

File: templates/components/api-server/component.js (hunk 6 above)
Comment (bajtos) [inline templates/components/api-server/component.js]: Actually I can take care of that myself, to speed things up.

File: templates/components/api-server/template/server/server.js (hunk 3 above)
Comment (bajtos) [inline templates/components/api-server/template/server/server.js]: > Maybe add a test for async boot scripts?

That would be great. Could you contribute it yourself please?

Let's do it in a standalone PR though.

**Comment by bajtos** [comment]
Landed via 18053ce, thank you for the contribution.

I had to also fix [server/server.js](https://github.com/strongloop/loopback-workspace/blob/18053cea59eac97503bd43b939ada6c50da086ac/server/server.js#L57-L61) to correctly load the new explorer component.

**Comment by bajtos** [comment]
Released in `loopback-workspace@3.18.0`.
